### ISSUE TITLE
build.sh example for go test -fuzz

### ISSUE THREAD

**Issue by hickford** [google/clusterfuzzlite#102]
I write fuzz tests in Go 1.18 following https://go.dev/doc/fuzz/ and run them with a command such as `go test -fuzz=Range`. How can I integrate them with ClusterFuzzLite?

I ran `python3 infra/helper.py generate --external --language=go $PATH_TO_PROJECT` but now I'm stuck. What do I put in build.sh? It would be neat to have an example in the documentation. The current example looks C++ specific https://google.github.io/clusterfuzzlite/build-integration/#buildsh

**Comment by hickford** [comment]
Ah reading https://google.github.io/clusterfuzzlite/build-integration/#prerequisites it looks like ClusterFuzzLite wants some other kind of fuzz target

> ClusterFuzzLite supports libFuzzer targets built with Clang on Linux.

**Comment by hickford** [comment]
Perhaps I can follow https://google.github.io/oss-fuzz/getting-started/new-project-guide/go-lang/#native-go-fuzzing-support though?

**Comment by jonathanmetzman** [comment]
@DavidKorczynski just confirming, oss-fuzz doesn't support go 1.18 fuzzing right?

**Comment by DavidKorczynski** [comment]
@AdamKorcz can you clarify details here?

My personal summary, although Adam knows a lot more about this: OSS-Fuzz supports fuzzing of targets written syntactically for go 1.18 native. These targets, however, are compiled such that they run by way of libFuzzer. A consequence of this is the corpus used for native fuzzing and the OSS-Fuzz style will be incompatible. i.e. you cannot upload your own corpus and reproduction of issues should happen by way of OSS-Fuzz. But besides that, OSS-Fuzz is fully compatible with go 1.18 targets and already larger projects are using it, e.g. Istio (https://github.com/google/oss-fuzz/pull/7992)

**Comment by DavidKorczynski** [comment]
@AdamKorcz could you perhaps make a sample go 1.18 ClusterFuzzLite integration if this would help?

**Comment by AdamKorcz** [comment]
Go 1.18 fuzz targets can run on OSS-Fuzz using `compile_native_go_fuzzer` which changes `&testing.F{}` to run with libFuzzer. As such, in the process of building Go 1.18 fuzzers on OSS-Fuzz they are made into libFuzzer binaries. All features in libFuzzer will subsequently be available whereas not all features in Go 1.18 fuzzing will.

**Comment by DavidKorczynski** [comment]
> Perhaps I can follow https://google.github.io/oss-fuzz/getting-started/new-project-guide/go-lang/#native-go-fuzzing-support though ? It mentions compile_native_go_fuzzer

This should work.

**Comment by DavidKorczynski** [comment]
> tl;dr no unless we can fuzz this target with libfuzzer.

And we should be able to do that with all go 1.18 fuzzers as long as they are compiled by way of `compile_native_go_fuzzer` (https://github.com/google/oss-fuzz/blob/master/infra/base-images/base-builder/compile_native_go_fuzzer)

**Comment by hickford** [comment]
Thanks I now have ClusterFuzzLite running in GitHub Actions https://github.com/hickford/merkle/pull/1

However when I tested it by deliberately breaking the test, the workflow succeeded regardless. Any ideas?

```
> go test -fuzz=Range
fuzz: elapsed: 0s, gathering baseline coverage: 0/42 completed
failure while testing seed corpus entry: FuzzRangeNodes/27d3f847ee8ebbcea67cc43e524feee604c2cfaf057db1c57a3d10b5bafd70cf
fuzz: elapsed: 0s, gathering baseline coverage: 0/42 completed
--- FAIL: FuzzRangeNodes (0.05s)
    --- FAIL: FuzzRangeNodes (0.00s)
        node_fuzz_test.go:16: begin=526, end=776
        node_fuzz_test.go:18: nodes=[{1 13} {2 7} {5 1} {6 1} {7 1} {8 1} {8 2} {3 96}]
        node_fuzz_test.go:25: got=26, want=526
    
FAIL
exit status 1
FAIL    github.com/transparency-dev/merkle/compact      2.868s
```

**Comment by AdamKorcz** [comment]
> Thanks I now have ClusterFuzzLite running in GitHub Actions [hickford/merkle#1](https://github.com/hickford/merkle/pull/1)
> 
> However when I tested it by deliberately breaking the test, the [workflow succeeded regardless](https://github.com/hickford/merkle/runs/7472433593?check_suite_focus=true). Any ideas?
> 
> Fuzz test fails locally
> 
> ```
> > go test -fuzz=Range
> fuzz: elapsed: 0s, gathering baseline coverage: 0/42 completed
> failure while testing seed corpus entry: FuzzRangeNodes/27d3f847ee8ebbcea67cc43e524feee604c2cfaf057db1c57a3d10b5bafd70cf
> fuzz: elapsed: 0s, gathering baseline coverage: 0/42 completed
> --- FAIL: FuzzRangeNodes (0.05s)
>     --- FAIL: FuzzRangeNodes (0.00s)
>         node_fuzz_test.go:16: begin=526, end=776
>         node_fuzz_test.go:18: nodes=[{1 13} {2 7} {5 1} {6 1} {7 1} {8 1} {8 2} {3 96}]
>         node_fuzz_test.go:25: got=26, want=526
>     
> FAIL
> exit status 1
> FAIL    github.com/transparency-dev/merkle/compact      2.868s
> ```

ClusterFuzzLite is not running the libFuzzer binary here. The fuzzer is being run by way of `go test -fuzz=`. 

Your workflow is correctly building the fuzzer with `compile_native_go_fuzzer github.com/transparency-dev/merkle/compact FuzzRangeNodes FuzzRangeNodes`. This produces the `FuzzRangeNodes` binary which must run in ClusterFuzzLite.

**Comment by AdamKorcz** [comment]
> Thanks I now have ClusterFuzzLite running in GitHub Actions [hickford/merkle#1](https://github.com/hickford/merkle/pull/1)
> 
> However when I tested it by deliberately breaking the test, the [workflow succeeded regardless](https://github.com/hickford/merkle/runs/7472433593?check_suite_focus=true). Any ideas?
> 
> Fuzz test fails locally as expected
> 
> ```
> > go test -fuzz=Range
> fuzz: elapsed: 0s, gathering baseline coverage: 0/42 completed
> failure while testing seed corpus entry: FuzzRangeNodes/27d3f847ee8ebbcea67cc43e524feee604c2cfaf057db1c57a3d10b5bafd70cf
> fuzz: elapsed: 0s, gathering baseline coverage: 0/42 completed
> --- FAIL: FuzzRangeNodes (0.05s)
>     --- FAIL: FuzzRangeNodes (0.00s)
>         node_fuzz_test.go:16: begin=526, end=776
>         node_fuzz_test.go:18: nodes=[{1 13} {2 7} {5 1} {6 1} {7 1} {8 1} {8 2} {3 96}]
>         node_fuzz_test.go:25: got=26, want=526
>     
> FAIL
> exit status 1
> FAIL    github.com/transparency-dev/merkle/compact      2.868s
> ```

The fuzzer should find this bug. Could you try increasing the fuzz time to 30 minutes?

**Comment by AdamKorcz** [comment]
Okay, I see the issue here. I will get a fix ready for this.

**Comment by AdamKorcz** [comment]
@hickford This should now have been fixed in https://github.com/AdamKorcz/go-118-fuzz-build. Could you trigger the CI job again and ensure that you are using the latest version of `github.com/AdamKorcz/go-118-fuzz-build`?

**Comment by hickford** [comment]
@AdamKorcz  Thanks, yes working as expected now. https://github.com/hickford/merkle/pull/1 When I try to break the test, the workflow fails. https://github.com/hickford/merkle/actions/runs/2732014085

**Comment by AdamKorcz** [comment]
> @AdamKorcz Thanks, yes working as expected now. [hickford/merkle#1](https://github.com/hickford/merkle/pull/1) When I try to break the test, the workflow fails. https://github.com/hickford/merkle/actions/runs/2732014085

That is great!

@jonathanmetzman this issue can be closed.
